src/common/libkvs/kvs_checkpoint.c

```diff
@@ -0,0 +1,132 @@
+/************************************************************\
+ * Copyright 2022 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <jansson.h>
+#include <flux/core.h>
+#include <time.h>
+
+#include "kvs_checkpoint.h"
+
+flux_future_t *kvs_checkpoint_commit (flux_t *h,
+                                      const char *key,
+                                      const char *rootref)
+{
+    flux_future_t *f = NULL;
+    double timestamp;
+
+    if (!h || !key || !rootref) {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    timestamp = flux_reactor_now (flux_get_reactor (h));
+
+    if (!(f = flux_rpc_pack (h,
+                             "kvs-checkpoint.put",
+                             0,
+                             0,
+                             "{s:s s:{s:i s:s s:f}}",
+                             "key",
+                             key,
+                             "value",
+                             "version", 1,
+                             "rootref", rootref,
+                             "timestamp", timestamp)))
+        return NULL;
+
+    return f;
+}
+
+flux_future_t *kvs_checkpoint_lookup (flux_t *h, const char *key)
+{
+    if (!h || !key) {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    return flux_rpc_pack (h,
+                          "kvs-checkpoint.get",
+                          0,
+                          0,
+                          "{s:s}",
+                          "key",
+                          key);
+}
+
+int kvs_checkpoint_lookup_get_rootref (flux_future_t *f, const char **rootref)
+{
+    const char *tmp_rootref;
+    int version;
+
+    if (!f || !rootref) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    if (flux_rpc_get_unpack (f, "{s:{s:i s:s}}",
+                                "value",
+                                "version", &version,
+                                "rootref", &tmp_rootref) < 0)
+        return -1;
+
+    if (version != 0 && version != 1) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    (*rootref) = tmp_rootref;
+    return 0;
+}
+
+/* returns "N/A" if not available */
+int kvs_checkpoint_lookup_get_formatted_timestamp (flux_future_t *f,
+                                                   char *buf,
+                                                   size_t len)
+{
+    int version;
+    double timestamp;
+
+    if (!f || !buf || len <= 0) {
```

Comment:
I don't think len can be less than zero as `size_t` is unsigned.

Reply:
Oh, actually, probably the right thing is to ignore `len` since zero isn't the only value that's too short.  Instead, check the return value of `snprintf()` and `strftime()` to catch any overflow.